File: string_to_integer/string_to_integer.py
#!/usr/bin/env python3
# -*- coding : utf-8 -*-

import re
class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        str = str.strip()
        m = re.match(r'^([-+]?[0-9]+).*', str)
        if not m:
            return 0
        num_of_str = m.group(1)
        num = int(num_of_str)
        if num < 0 and num < (-2 ** 31):
            return -2**31
        if num > 0 and num > 2**31 - 1:
            return 2**31 - 1
        return num

if __name__ == '__main__':
    s = Solution()
    print(s.myAtoi("42"))
    print(s.myAtoi(""))
    print(s.myAtoi("     -42"))
    print(s.myAtoi("4223 with word"))
    print(s.myAtoi("word with 42"))
    print(s.myAtoi("-912834723320000000"))
    print(s.myAtoi("91283472332"))

File: string_to_integer/string_to_integer.cpp
#include<iostream>
#include<string>
#include<limits.h>
using namespace std;
class Solution {
public:
    int myAtoi(string str) {
        int len = str.size();
        if(str.size() == 0){
            return 0;
        }
        int i = 0;
        while(i<len && str[i] == ' ')i++;
        if(i>=len)return 0;
        int sign = 1;
        if(str[i] == '-'){
            sign = -1;
            i++;
        } else if(str[i]=='+'){
            i++;
        }
        if(str[i]<'0' || str[i]>'9'){
            return 0;
        }
        int num = 0;
        while(i<len && str[i]>='0' && str[i]<='9'){
            int add=str[i]-'0';
            //cout<<str<<' '<<add<<endl;
            if(sign == 1){
                if((INT_MAX-add)/10.0<num)return INT_MAX;
                //cout<(INT_MAX-add)/num<<endl;
            } else {
                if((-INT_MIN+add)/10.0>-num)return INT_MIN;
            }
            num = num*10 + add;
            i++;
        }
        return sign*num;
    }
};
int main(){
    Solution s;
    cout<<s.myAtoi("42")<<endl;
    cout<<s.myAtoi("   -42")<<endl;
    cout<<s.myAtoi("42 with world")<<endl;
    cout<<s.myAtoi("with world 42")<<endl;
    cout<<s.myAtoi("-91283472332")<<endl;

    return 0;
}